using_restricted_array.rb

```diff
@@ -7,41 +7,96 @@
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
 def length(array)
-  puts "NOT IMPLEMENTED"
+  length = 0
+  while !array[length].nil?
+    length += 1
+  end
+  return length
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  while i<length(array)
+    print "#{array[i]} "
+    i+=1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  while i < length
+    if array[i] == value_to_find
+      return true
+    end
+    i+=1
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  largest = array[0]
+  i = 1
+  while i < length
+    array[i] > largest ? largest = array[i] : largest
+    i+=1
+  end
+  return largest
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+  smallest_num = array[0]
+  i=1
+  while i<length
+    array[i] < smallest_num ? smallest_num = array[i] : smallest_num
+    i+=1
+  end
+  return smallest_num
 end
 
 # Reverses the values in the integer array in place
 def reverse(array, length)
-  puts "NOT IMPLEMENTED"
+  j = length-1
+  i = 0
+  until i >= j
+    temp = array[j]
+    array[j] = array[i]
+    array[i] = temp
+    i+=1
+    j-=1
+  end
+  return array
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def binary_search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  min = 0
```

Comment:
Please take a look at the updated binary_search slide. 
Here's the considerations I was missing in the algorithm on the slide before the update: (1) what if there is only one element in the array and it is the value_to_find? In this case, it used to return false. (To fix this, add an additional if statement after the loop ends, to check if array[min] is the value_to_find.) (2) Consider the scenario where there are two elements in the array and the value_to_find is the second element in the array. In the first iteration, min = 0, max=1, mid=0. Value is not at index 0. So, min = mid (still 0). On re-entering the loop, mid would be still 0. So we had an infinite loop. (To fix this, update max to be mid -1, if value_to_find is less than mid. And update min = mid+1, if value_to_find is greater than mid.)